lib/puppet/type/file.rb

```diff
@@ -887,6 +887,31 @@ def write(property = nil)
 
   private
 
+  def set_sensitive_properties(sensitive_parameters)
+    # The source parameter isn't actually a property but works by injecting information into the
+    # content property. In order to preserve the intended sensitive context we need to mark content
+    # as sensitive instead.
+    if sensitive_parameters.include?(:source)
+      sensitive_parameters.delete(:source)
+      sensitive_parameters << :content
+      # The `source` parameter will generate the `content` property when the resource state is retrieved
+      # but that's long after we've set the sensitive context. Force the early creation of the `content`
+      # attribute so we can mark it as sensitive.
+      newattr(:content)
+    end
+
+    # If we have content that's marked as sensitive but the file doesn't exist then the ensure
+    # property will be responsible for syncing content, so we have to mark ensure as sensitive as well.
+    if sensitive_parameters.include?(:content)
```

Comment:
This is only necessary if the file doesn't exist and might unnecessarily redact information for files that exist. That being said I think this'll only come up in bizarre edge cases so we might be able to wing it.